src/expansions.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright 2021-2023 SecureDNA Stiftung (SecureDNA Foundation) <licensing@securedna.org>
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::sync::Arc;
+
+use smallvec::SmallVec;
+
+use crate::{
+    BaseSequence, DnaSequenceAmbiguous, DnaSequenceStrict, Nucleotide, NucleotideAmbiguous,
+};
+
+/// Iterator of all unambiguous expansions of ambiguous DNA.
+///
+/// Expansions are returned in lexicographic order based on the ordering of [`Nucleotide`]
+/// (not currently alphabetical).
+#[derive(Clone)]
+pub struct Expansions {
+    ambiguities: SmallVec<[Ambiguity; 8]>,
+    buf: Arc<[Nucleotide]>,
+    began: bool,
+}
+
+#[derive(Clone)]
+struct Ambiguity {
+    index: usize,
+    digit: u8,
+    nucleotide: NucleotideAmbiguous,
+}
+
+/// Unambiguous DNA expansion produced by [`Expansions`] iterator.
+#[derive(Clone, PartialEq, Eq)]
+pub struct Expansion(Arc<[Nucleotide]>);
+
+impl Expansions {
+    // Construct new [`Expansions`] iterator
+    pub fn new(dna: &[NucleotideAmbiguous]) -> Self {
+        let ambiguities = dna
+            .iter()
+            .enumerate()
+            .filter(|(_, nuc)| nuc.possibilities().len() > 1)
+            .map(|(index, &nucleotide)| Ambiguity {
+                index,
+                digit: 0,
+                nucleotide,
+            })
+            .collect();
+        let buf: SmallVec<[_; 64]> = dna
+            .iter()
+            .map(|nuc| *nuc.possibilities().first().unwrap())
+            .collect();
+        let buf = Arc::from(buf.as_slice());
+        Expansions {
+            ambiguities,
+            buf,
+            began: false,
+        }
+    }
+}
+
+impl Iterator for Expansions {
+    type Item = Expansion;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.began {
+            self.began = true;
+            return Some(Expansion(self.buf.clone()));
+        }
+
+        // Can't use Arc::make_mut because [T] is unsized
+        let buf = match Arc::get_mut(&mut self.buf) {
+            Some(buf) => buf,
+            None => {
+                self.buf = Arc::from(&*self.buf);
+                Arc::get_mut(&mut self.buf).unwrap()
+            }
+        };
+
+        for amb in self.ambiguities.iter_mut().rev() {
+            amb.digit = (amb.digit + 1) % (amb.nucleotide.possibilities().len() as u8);
```

Comment:
I wonder if it makes sense (and if it even makes a performance difference) to make a method on nucleotides that immediately returns the number of possibilities, rather than write `.possibilities().len()`.

Also I really like this algorithm :)

Reply:
I'm thinking the implementation can be `usize::count_ones(amb.nucleotide.into())`, which is often a single operation (`popcnt` in x86).

Reply:
Oh! That's an even better idea. I'll try that.

Reply:
I tried this out on my laptop with no noticeable difference, so I was about to discard that change... but just to be diligent I tried it on `biofilter` and it consistently yielded a whopping 15-20% performance improvement! I really need to be better about doing performance testing on `biofilter` instead of my dev machine.